module/campaign/gems_farming.py

```diff
@@ -22,6 +28,48 @@
 SIM_VALUE = 0.92
 
 
+class GemsEmotion(Emotion):
+
+    def check_reduce(self, battle):
+        """
+        Overwrite emotion.check_reduce()
+        Check emotion before entering a campaign.
+
+        Args:
+            battle (int): Battles in this campaign
+
+        Raise:
+            CampaignEnd: Pause current task to prevent emotion control in the future.
+        """
+        
+        if not self.is_calculate:
+            return
+
+        method = self.config.Fleet_FleetOrder
+
+        if method == 'fleet1_all_fleet2_standby':
+            battle = (battle, 0)
+        elif method == 'fleet1_standby_fleet2_all':
+            battle = (0, battle)
+        else:
+            raise ScriptError(f'Unknown fleet order: {method}')
+
+        battle = tuple(np.array(battle) * self.reduce_per_battle_before_entering)
+        logger.info(f'Expect emotion reduce: {battle}')
+
+        self.update()
+        self.record()
+        self.show()
```

Comment:
尽量复用父类的 check_reduce，而不是完全重写

Reply:
父类check_reduce里面有一个config.task_delay(target=recovered)，如何比较优雅地剔除这个带来的影响？
我想到的一种比较屎的办法是
```
try:
  super().check_reduce(self, battle)
except ScriptEnd:
  self.config.GEMS_EMOTION_TRIGGERED = True
  self.config.task_call("GemsFarming") # to undo emotion delay
  raise CampaignEnd('Emotion control')
```
不知道你怎么看

Reply:
鉴定为屎方法：
```
─────────────────────────────────────── B1 ────────────────────────────────────────
INFO     12:37:24.813 │ B1                                                         
INFO     12:37:24.814 │ Expect emotion reduce: (10, 0)                             
INFO     12:37:24.838 │ Bind task ['General', 'Alas', 'TaskBalancer',              
         'EventGeneral', 'GemsFarming']                                            
INFO     12:37:24.840 │ Save config ./config\alas.json,                            
         GemsFarming.Emotion.Fleet1Value=8,                                        
         GemsFarming.Emotion.Fleet1Record=datetime.datetime(2024, 8, 27, 12, 37,   
         24), GemsFarming.Emotion.Fleet2Value=119,                                 
         GemsFarming.Emotion.Fleet2Record=datetime.datetime(2024, 8, 27, 12, 37,   
         24)                                                                       
INFO     12:37:24.850 │ [Emotion fleet_1] 8                                        
INFO     12:37:24.852 │ [Emotion fleet_2] 119                                      
INFO     12:37:24.853 │ Delay current task to prevent emotion control in the future
INFO     12:37:24.855 │ Delay task `GemsFarming` to 2024-08-27 12:54:00            
         (target=datetime.datetime(2024, 8, 27, 12, 54))                           
INFO     12:37:24.879 │ Bind task ['General', 'Alas', 'TaskBalancer',              
         'EventGeneral', 'GemsFarming']                                            
INFO     12:37:24.881 │ Save config ./config\alas.json,                            
         GemsFarming.Scheduler.NextRun=datetime.datetime(2024, 8, 27, 12, 54)      
INFO     12:37:24.889 │ Task call: GemsFarming                                     
INFO     12:37:24.912 │ Bind task ['General', 'Alas', 'TaskBalancer',              
         'EventGeneral', 'GemsFarming']                                            
INFO     12:37:24.914 │ Save config ./config\alas.json,                            
         GemsFarming.Scheduler.NextRun=datetime.datetime(2024, 8, 27, 12, 37, 24), 
         GemsFarming.Scheduler.Enable=True                                         
═════════════════════════════════ CHANGE FLAGSHIP ═════════════════════════════════
INFO     12:37:24.924 │ CHANGE FLAGSHIP                                            
INFO     12:37:24.925 │ [ChangeFlagship] ship_equip                                
INFO     12:37:24.926 │ <<< UI ENSURE >>>                                          
INFO     12:37:24.927 │ UI get current page                                        
INFO     12:37:24.934 │ [UI] page_event                                            
INFO     12:37:24.936 │ Goto page_fleet                                            
INFO     12:37:24.938 │ <<< UI GOTO PAGE_FLEET >>>                                 
INFO     12:37:24.948 │ Page switch: page_event -> page_main                       
INFO     12:37:24.949 │ Click (1242,   33) @ GOTO_MAIN                             
INFO     12:37:25.822 │ Page switch: page_main_white -> page_fleet                 
INFO     12:37:25.824 │ Click (1064,  482) @ MAIN_GOTO_FLEET_WHITE                 
INFO     12:37:26.823 │ Page arrive: page_fleet                                    
INFO     12:37:26.825 │ <<< UI ENSURE INDEX >>>                                    
INFO     12:37:26.837 │ [OCR_FLEET_INDEX 0.011s] 1                                 
INFO     12:37:26.838 │ [Index] 1                                                  
INFO     12:37:27.074 │ Click (1250,  347) @ FLEET_NEXT                            
INFO     12:37:27.291 │ Click (1244,  337) @ FLEET_NEXT                            
INFO     12:37:27.363 │ [OCR_FLEET_INDEX 0.007s] 2                                 
INFO     12:37:27.364 │ [Index] 2                                                  
INFO     12:37:27.594 │ [OCR_FLEET_INDEX 0.006s] 2                                 
INFO     12:37:27.596 │ [Index] 2                                                  
INFO     12:37:27.765 │ [OCR_FLEET_INDEX 0.007s] 2                                 
INFO     12:37:27.766 │ [Index] 2                                                  
INFO     12:37:27.964 │ [OCR_FLEET_INDEX 0.008s] 2                                 
INFO     12:37:27.965 │ [Index] 2                                                  
INFO     12:37:28.164 │ [OCR_FLEET_INDEX 0.007s] 2                                 
INFO     12:37:28.165 │ [Index] 2                                                  
INFO     12:37:28.368 │ [OCR_FLEET_INDEX 0.008s] 2                                 
INFO     12:37:28.370 │ [Index] 2                                                  
INFO     12:37:28.653 │ Click (1253,  342) @ FLEET_NEXT                            
INFO     12:37:28.723 │ [OCR_FLEET_INDEX 0.007s] 2                                 
INFO     12:37:28.724 │ [Index] 2                                                  
INFO     12:37:28.927 │ [OCR_FLEET_INDEX 0.009s] 3                                 
INFO     12:37:28.928 │ [Index] 3                                                  
─────────────────────────── UNMOUNT FLAGSHIP EQUIPMENTS ───────────────────────────
INFO     12:37:28.930 │ UNMOUNT FLAGSHIP EQUIPMENTS                                
INFO     12:37:28.931 │ <<< UI CLICK >>>                                           
INFO     12:37:28.941 │ Click ( 948,  679) @ FLEET_DETAIL                          
INFO     12:37:29.330 │ Click ( 131,  404) @ FLEET_DETAIL_ENTER_FLAGSHIP           
INFO     12:37:29.922 │ Click ( 987,  100) @ EQUIPMENT_CODE_ENTRANCE               
INFO     12:37:30.643 │ Click (1063,  507) @ EQUIPMENT_CODE_CLEAR                  
INFO     12:37:30.928 │ Confirm equipment preview cleared.                         
INFO     12:37:30.930 │ Click (1055,  576) @ EQUIPMENT_CODE_CONFIRM                
INFO     12:37:31.333 │ Click ( 778,  513) @ POPUP_CONFIRM_EQUIPMENT_CODE          
INFO     12:37:32.942 │ <<< UI CLICK >>>                                           
INFO     12:37:33.144 │ Click (  47,   44) @ BACK_ARROW                            
INFO     12:37:34.151 │ <<< UI CLICK >>>                                           
INFO     12:37:34.354 │ Click (  60,   56) @ BACK_ARROW                            
───────────────────────────────── CHANGE FLAGSHIP ─────────────────────────────────
INFO     12:37:34.756 │ CHANGE FLAGSHIP                                            
INFO     12:37:34.757 │ <<< UI CLICK >>>                                           
INFO     12:37:34.768 │ Click ( 594,  285) @ FLEET_ENTER_FLAGSHIP                  
INFO     12:37:35.356 │ <<< UI CLICK >>>                                           
INFO     12:37:35.365 │ Click (1148,   26) @ DOCK_FILTER                           
INFO     12:37:36.162 │ Setting options DOCK,                                      
INFO     12:37:36.171 │ [DOCK] sort/level, index/all, faction/all, rarity/all,     
         extra/no_limit                                                            
INFO     12:37:36.178 │ Setting options DOCK, sort='total', index='cv',            
         faction='all', rarity='common', extra='enhanceable'                       
INFO     12:37:36.184 │ [DOCK] sort/level, index/all, faction/all, rarity/all,     
         extra/no_limit                                                            
INFO     12:37:36.191 │ Click ( 564,   88) @ FILTER_SORT_2_0                       
INFO     12:37:36.243 │ Click ( 312,  217) @ FILTER_INDEX_0_1                      
INFO     12:37:36.296 │ Click ( 462,  415) @ FILTER_RARITY_1_0                     
INFO     12:37:36.350 │ Click ( 696,  500) @ FILTER_EXTRA_3_0                      
INFO     12:37:36.419 │ [DOCK] sort/total, index/all, faction/all, rarity/all,     
         extra/no_limit                                                            
INFO     12:37:36.619 │ [DOCK] sort/total, index/cv, faction/all, rarity/common,   
         extra/no_limit                                                            
INFO     12:37:36.820 │ [DOCK] sort/total, index/cv, faction/all, rarity/common,   
         extra/no_limit                                                            
INFO     12:37:37.021 │ [DOCK] sort/total, index/cv, faction/all, rarity/common,   
         extra/no_limit                                                            
INFO     12:37:37.221 │ [DOCK] sort/total, index/cv, faction/all, rarity/common,   
         extra/no_limit                                                            
INFO     12:37:37.424 │ [DOCK] sort/total, index/cv, faction/all, rarity/common,   
         extra/no_limit                                                            
INFO     12:37:37.432 │ Click ( 724,  496) @ FILTER_EXTRA_3_0                      
INFO     12:37:37.623 │ [DOCK] sort/total, index/cv, faction/all, rarity/common,   
         extra/enhanceable                                                         
INFO     12:37:37.630 │ <<< UI CLICK >>>                                           
INFO     12:37:37.633 │ Click ( 807,  637) @ DOCK_FILTER_CONFIRM                   
INFO     12:37:39.188 │ [Favourite_filter] off                                     
INFO     12:37:39.190 │ <<< FINDING FLAGSHIP >>>                                   
INFO     12:37:39.192 │ Limitaions set to {'level': (1, 31), 'emotion': (10, 150), 
         'rarity': 'any', 'fleet': 3, 'status': 'free'}                            
INFO     12:37:39.194 │ [Dork_sorting] Descending                                  
INFO     12:37:39.195 │ Click (1034,   33) @ SORTING_CLICK                         
INFO     12:37:39.394 │ [Dork_sorting] Ascending                                   
INFO     12:37:40.830 │ [DOCK_LEVEL_OCR 0.117s] [0, 30, 28, 70, 0, 0, 0, 0, 0, 0,  
         0, 0, 0, 0]                                                               
INFO     12:37:40.940 │ [DOCK_EMOTION_OCR 0.107s] [0, 8, 8, 119, 0, 0, 0, 0, 0, 0, 
         0, 0, 0, 0]                                                               
INFO     12:37:41.014 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_0_0)                                           
INFO     12:37:41.016 │ Ship(rarity=None, level=30, emotion=8, fleet=3,            
         status='free', button=CARD_1_0)                                           
INFO     12:37:41.017 │ Ship(rarity=None, level=28, emotion=8, fleet=0,            
         status='free', button=CARD_2_0)                                           
INFO     12:37:41.018 │ Ship(rarity=None, level=70, emotion=119, fleet=0,          
         status='free', button=CARD_3_0)                                           
INFO     12:37:41.019 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_4_0)                                           
INFO     12:37:41.020 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_5_0)                                           
INFO     12:37:41.021 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_6_0)                                           
INFO     12:37:41.023 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_0_1)                                           
INFO     12:37:41.024 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_1_1)                                           
INFO     12:37:41.025 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_2_1)                                           
INFO     12:37:41.026 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_3_1)                                           
INFO     12:37:41.027 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_4_1)                                           
INFO     12:37:41.028 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_5_1)                                           
INFO     12:37:41.030 │ Ship(rarity=None, level=0, emotion=0, fleet=0,             
         status='free', button=CARD_6_1)                                           
INFO     12:37:41.031 │ Limitaions set to {'level': (1, 31), 'emotion': (10, 150), 
         'rarity': 'any', 'fleet': 0, 'status': 'free'}                            
INFO     12:37:41.148 │ [DOCK_LEVEL_OCR 0.116s] [0, 30, 28, 70, 0, 0, 0, 0, 0, 0,  
         0, 0, 0, 0]                                                               
INFO     12:37:41.249 │ [DOCK_EMOTION_OCR 0.099s] [0, 8, 8, 119, 0, 0, 0, 0, 0, 0, 
         0, 0, 0, 0]                                                               
INFO     12:37:41.323 │ Change flagship failed, no CV in common rarity.            
INFO     12:37:41.325 │ <<< UI CLICK >>>                                           
INFO     12:37:41.329 │ Click (1135,   25) @ DOCK_FILTER                           
INFO     12:37:41.795 │ Setting options DOCK,                                      
INFO     12:37:41.803 │ [DOCK] sort/total, index/cv, faction/all, rarity/common,   
         extra/enhanceable                                                         
INFO     12:37:41.809 │ Click ( 469,   89) @ FILTER_SORT_1_0                       
INFO     12:37:41.862 │ Click ( 313,  157) @ FILTER_INDEX_0_0                      
INFO     12:37:41.915 │ Click ( 302,  419) @ FILTER_RARITY_0_0                     
INFO     12:37:41.968 │ Click ( 317,  487) @ FILTER_EXTRA_0_0                      
INFO     12:37:42.036 │ [DOCK] sort/level, index/all, faction/all, rarity/common,  
         extra/enhanceable                                                         
INFO     12:37:42.240 │ [DOCK] sort/level, index/all, faction/all, rarity/all,     
         extra/no_limit                                                            
INFO     12:37:42.248 │ Setting options DOCK, sort='level', index='all',           
         faction='all', rarity='all', extra='no_limit'                             
INFO     12:37:42.255 │ [DOCK] sort/level, index/all, faction/all, rarity/all,     
         extra/no_limit                                                            
INFO     12:37:42.261 │ <<< UI CLICK >>>                                           
INFO     12:37:42.264 │ Click ( 786,  646) @ DOCK_FILTER_CONFIRM                   
INFO     12:37:43.792 │ [Favourite_filter] off                                     
INFO     12:37:43.793 │ [Dork_sorting] Ascending                                   
INFO     12:37:43.794 │ Click (1023,   25) @ SORTING_CLICK                         
INFO     12:37:43.990 │ [Dork_sorting] unknown                                     
INFO     12:37:44.192 │ [Dork_sorting] Descending                                  
INFO     12:37:45.541 │ <<< UI CLICK >>>                                           
INFO     12:37:45.742 │ Click (  50,   63) @ BACK_ARROW                            
──────────────────────────── MOUNT FLAGSHIP EQUIPMENTS ────────────────────────────
INFO     12:37:46.548 │ MOUNT FLAGSHIP EQUIPMENTS                                  
INFO     12:37:46.550 │ <<< UI CLICK >>>                                           
INFO     12:37:46.553 │ Click ( 955,  657) @ FLEET_DETAIL                          
INFO     12:37:46.948 │ Click ( 121,  388) @ FLEET_DETAIL_ENTER_FLAGSHIP           
INFO     12:37:47.553 │ Click ( 984,   95) @ EQUIPMENT_CODE_ENTRANCE               
INFO     12:37:48.155 │ Confirm equipment preview cleared.                         
INFO     12:37:48.182 │ Apply gear code SDBNL0lNRS9QVE0vMTRJLzE0SVww for bogue     
INFO     12:37:48.184 │ Click ( 731,  686) @ EQUIPMENT_CODE_TEXTBOX                
INFO     12:37:48.557 │ Click ( 944,  685) @ EQUIPMENT_CODE_ENTER                  
INFO     12:37:48.959 │ Confirm gear code loaded.                                  
INFO     12:37:48.960 │ Click (1031,  563) @ EQUIPMENT_CODE_CONFIRM                
INFO     12:37:49.364 │ Click ( 775,  511) @ POPUP_CONFIRM_EQUIPMENT_CODE          
INFO     12:37:50.572 │ Gear code import complete.                                 
INFO     12:37:50.573 │ <<< UI CLICK >>>                                           
INFO     12:37:50.774 │ Click (  60,   48) @ BACK_ARROW                            
INFO     12:37:51.780 │ <<< UI CLICK >>>                                           
INFO     12:37:51.984 │ Click (  58,   74) @ BACK_ARROW                            
INFO     12:37:52.197 │ Pending tasks: ['GemsFarming', 'OpsiHazard1Leveling']      
INFO     12:37:52.198 │ [Task] GemsFarming (Enable, 2024-08-27 12:37:24)           
INFO     12:37:52.200 │ Switch task `GemsFarming (Enable, 2024-08-27 12:36:53)` to 
         `GemsFarming (Enable, 2024-08-27 12:37:24)`                               
INFO     12:37:52.202 │ Scheduler: End task `GemsFarming`                          
INFO     12:37:52.203 │ [Server] jp                                                
INFO     12:37:52.217 │ Bind task ['General', 'Alas']                              
INFO     12:37:52.220 │ Pending tasks: ['GemsFarming', 'OpsiHazard1Leveling']      
INFO     12:37:52.221 │ [Task] GemsFarming (Enable, 2024-08-27 12:37:24)           
INFO     12:37:52.223 │ Bind task ['General', 'Alas', 'TaskBalancer',              
         'EventGeneral', 'GemsFarming']                                            
INFO     12:37:52.233 │ Scheduler: Start task `GemsFarming`                        
```
由于task同名不同时会被认为是两个不同任务，所以task_call会导致在卡船的时候寄。我的建议还是重写，因为重构父类check_reduce的代价太大。

Reply:
把task_call改成task_delay(minutes=-30)了。但是还是丑。

Reply:
参考群友建议把check_reduce拆成两部分了